go/obscuronode/enclave/rollupchain/rollup_chain.go

```diff
@@ -0,0 +1,699 @@
+	paramBytes, err := rc.rpcEncryptionManager.DecryptRPCCall(encryptedParams)
+	if err != nil {
+		return nil, fmt.Errorf("could not decrypt params in eth_call request. Cause: %w", err)
+	}
+
+	contractAddress, from, data, err := extractCallParams(paramBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	hs := rc.storage.FetchHeadState()
+	if hs == nil {
+		panic("Not initialised")
+	}
+	// todo - get the parent
+	r, f := rc.storage.FetchRollup(hs.HeadRollup)
+	if !f {
+		panic("not found")
+	}
+	s := rc.storage.CreateStateDB(hs.HeadRollup)
+	result, err := evm.ExecuteOffChainCall(from, contractAddress, data, s, r.Header, rc.storage, rc.obscuroChainID)
+	if err != nil {
+		return nil, err
+	}
+	if result.Failed() {
+		log.Info("Failed to execute contract %s: %s\n", contractAddress.Hex(), result.Err)
+		return nil, result.Err
+	}
+
+	encryptedResult, err := rc.rpcEncryptionManager.EncryptWithViewingKey(from, result.ReturnData)
+	if err != nil {
+		return nil, fmt.Errorf("enclave could not respond securely to eth_call request. Cause: %w", err)
+	}
+
+	return encryptedResult, nil
+}
+
+// Extracts and validates the relevant parameters in a Call request.
+func extractCallParams(decryptedParams []byte) (common.Address, common.Address, []byte, error) {
+	var paramsJSONMap []interface{}
+	err := json.Unmarshal(decryptedParams, &paramsJSONMap)
+	if err != nil {
+		return common.Address{}, common.Address{}, nil, fmt.Errorf("could not parse JSON params in Call request. Cause: %w", err)
+	}
+
+	txArgs := paramsJSONMap[0] // The first argument is the transaction arguments, the second the block, the third the state overrides.
+	contractAddressString, ok := txArgs.(map[string]interface{})[CallFieldTo].(string)
+	if !ok {
+		return common.Address{}, common.Address{}, nil, fmt.Errorf("to field in Call request params was not of expected type string")
+	}
+	fromString, ok := txArgs.(map[string]interface{})[CallFieldFrom].(string)
+	if !ok {
+		return common.Address{}, common.Address{}, nil, fmt.Errorf("from field in Call request params was not of expected type string")
+	}
+	dataString, ok := txArgs.(map[string]interface{})[CallFieldData].(string)
+	if !ok {
+		return common.Address{}, common.Address{}, nil, fmt.Errorf("data field in Call request params was not of expected type string")
+	}
+
+	contractAddress := common.HexToAddress(contractAddressString)
+	from := common.HexToAddress(fromString)
+	data, err := hexutil.Decode(dataString)
+	if err != nil {
+		return common.Address{}, common.Address{}, nil, fmt.Errorf("could not decode data in Call request. Cause: %w", err)
+	}
+	return contractAddress, from, data, nil
+}
+
+func (rc *RollupChain) GetBalance(encryptedParams nodecommon.EncryptedParamsGetBalance) (nodecommon.EncryptedResponseGetBalance, error) {
+	paramBytes, err := rc.rpcEncryptionManager.DecryptRPCCall(encryptedParams)
```

Comment:
I wonder if it might be a cleaner separation if we could keep rpcEncryptionManager layer out of this rollup_chain logic. So if the enclave dealt with decrypting/encrypting and then just delegated the request to rollup_chain with the actual params.

Reply:
totally. I had the exact same thought. I even had a comment at some point in the struct to remove it.
If I had done this as part of this PR it would have violated the prime directive, as it would have been a bit more elaborate to extract it.